### ISSUE TITLE
Loom: Investigate performance impacts of pinning

### ISSUE THREAD

**Issue by tajila** [eclipse-openj9/openj9#15185]
Any operation that pins is essentially going to be serialized with virtual threads. This can severely degrade performance if the pinned operation is on the critical path. Note, any native function entry or code that runs in serialized block will pin. OpenJDK has addressed this in reflection (JEP 416), internet resolution (JEP 418), security managers (JEP 411) and other areas. We need to take a look at the core JCL to find places where we have native calls or synchronized blocks in hot paths and mitigate those with explicit locks or other mechanisms.

**Comment by tajila** [comment]
FYI @0xdaryl @vijaysun-omr

**Comment by vijaysun-omr** [comment]
Similar typo question about : "to find places where we have native calls or synchronized blocks in hot paths" should be "to find places where we have native calls and synchronized blocks (both) in hot paths" ?

**Comment by vijaysun-omr** [comment]
One clarification regarding your mention of the phrase "serialization" : "synchronization" anyway is a "serializing operation" (for the threads at the critical section) in a sense for the threads involved. So I wanted to better understand what additional "serialization" would be the case here with virtual threads. e.g. does it affect the carrier thread associated with the virtual thread only OR somehow "all" threads ?

**Comment by vijaysun-omr** [comment]
I'll ask about the example I had in mind wrt my interpretation of what you wrote : "Imagine we had 100 virtual threads and 20 carrier threads that they could run on...now imagine that 20 of those threads ended up doing a native call inside a sync region, meaning they bring the system to a standstill since the 80 other threads no longer have a carrier thread to ride on...if instead there were 100 real threads, then it doesn't matter what those 20 threads do, the other 80 threads can still get access to the CPU". Is this the right picture ?

**Comment by vijaysun-omr** [comment]
@zl-wang fyi too

**Comment by tajila** [comment]
> I'll ask about the example I had in mind wrt my interpretation of what you wrote : "Imagine we had 100 virtual threads and 20 carrier threads that they could run on...now imagine that 20 of those threads ended up doing a native call inside a sync region, meaning they bring the system to a standstill since the 80 other threads no longer have a carrier thread to ride on...if instead there were 100 real threads, then it doesn't matter what those 20 threads do, the other 80 threads can still get access to the CPU". Is this the right picture ?

Yes, thats correct. Basically, if a VirtualThread is pinned to the carrier (via `synchronized` or native code) and then blocks on something, then the carrier thread needs to wait for that blocking operation to complete, it cannot be used to schedule another thread. This effectively takes the carrier thread out of the pool of threads to schedule work. As stated in your example, if this happens to all Virtual threads mounted on carriers, then you are left with no carrier threads to do work until the blocking operations complete. 

> One clarification regarding your mention of the phrase "serialization" : "synchronization" anyway is a "serializing operation" (for the threads at the critical section) in a sense for the threads involved. So I wanted to better understand what additional "serialization" would be the case here with virtual threads. e.g. does it affect the carrier thread associated with the virtual thread only OR somehow "all" threads ?

So what I mean by serialization is that if a VirtualThread is pinned and blocks, then the other VirtualThreads (waiting to be scheduled) must wait until the pinned thread completes its blocking operation. It means that one cannot interleave the execution of a blocked VirtualThread with an active VirtualThread effectively serializing the execution of VirtualThreads.

**Comment by tajila** [comment]
> on the other hand, i took what you meant by pinning: any code snippet with blocking operations inside hasn't been re-wrote or transformed to use explicit continuations (i.e. allowing the carrier-thread go to the virtual-thread scheduler code).

So by "explicit locks" I mean java.util.concurrent.locks. These locks have been re-written such that they are yield points for VirtualThreads. Whenever a VirtualThread blocks on a j.u.c.lock, the scheduler can remove it and schedule another VirtualThread.

However, `synchronized` blocks do not have this support be cause the locking mechanism is internal to the JVM, making it difficult to integrate with the Java level scheduler. This limitation may be removed in the future, but for now it remains.

**Comment by 0xdaryl** [comment]
Just seeking some clarity as I suspect we're not all on the same page yet.  Based on my reading and interpretation of the wording here [1], pinning will occur if EITHER we enter a native OR a synchronized block or method (i.e., only ONE need be true).  Is that the correct interpretation?

[1] https://cr.openjdk.java.net/~rpressler/loom/loom/sol1_part1.html#pinning

**Comment by tajila** [comment]
> pinning will occur if EITHER we enter a native OR a synchronized block or method (i.e., only ONE need be true). Is that the correct interpretation?

Yes, thats correct. Only one needs to be true. 

On a related note, calling Object.wait() will pin as well.

**Comment by zl-wang** [comment]
> However, synchronized blocks do not have this support be cause the locking mechanism is internal to the JVM, making it difficult to integrate with the Java level scheduler. This limitation may be removed in the future, but for now it remains.

this sounds really ugly.  but, overall, this (pinning to the extent that it kind-of deadlocks ... i.e. other virtual-threads cannot make progress to un-pin the whole situation) is less a performance issue, but more a configuration/correctness issue.
